MarcusBarnes/mik
# Work on #363.

**Github issue**: (#363)

# What does this Pull Request do?

Adds a new metadata parser class that generates output from Twig templates.

# What's new?

A new metadata parser class that generates its output from Twig templates. It is a drop-in replacement for existing metadata parsers. This parser does not use field mappings; instead, source field values from the fetcher are interpolated directly into the templates.

# How should this be tested?

1. Check out the issue-363 branch.
1. Run `composer dump-autoload`.
1. Run PHPUnit tests: `phpunit --exclude-group inputvalidators --bootstrap vendor/autoload.php tests`. There should be no failures.
1. Unzip the attached configuration files and run mik: `./mik -c templated.ini`.
1. Look in the output directory. The XML files should correspond to the templated_mods_twig.xml template.
1. In the templated.ini file, comment out `template = templated_mods_twig.xml` and uncomment `; template = templated_dc_twig.xml`.
1. Delete your output directory.
1. Rerun mik.
1. Look in the output directory. The XML files should correspond to the templated_dc_twig.xml template.

To test the accompanying metadata manipulator:

1. Delete the MIK output and temp directories.
2. Add the following options to the .ini file:
```
[MANIPULATORS]
metadatamanipulators[] = "SimpleReplaceTemplated|/Island/|Peninsula"

[LOGGING]
path_to_manipulator_log = /tmp/templated_output/manipulator.log
```
3. Rerun MIK. 
4. In image02.xml and image05.xml, the word 'Island' will be replaced by the word 'Peninsula'.

# Additional Notes

We would need to document this metadata parser and accompanying manipulator. It should work with CONTENTdm and CSV toolchains. This metadata parser will also serve as a good example of how to extend MIK.

# Interested parties
@MarcusBarnes. Other MIK users may want to test this new parser class as well and offer feedback.

[templated.zip](https://github.com/MarcusBarnes/mik/files/871591/templated.zip)
